### PR TITLE
Add new model vSHARP & CMR challenge along with 3D reconstruction code

### PR DESCRIPTION
**Pull request by georgeyiasemis**
- **vSHARP Model**: Supporting both 2D and 3D. (Details: [vSHARP Paper](https://arxiv.org/abs/2309.09954))
- **CMRxRecon Challenge  Code**: Includes necessary code and dataset for the CMRxRecon challenge. Learn more about the challenge [here](https://cmrxrecon.github.io/Challenge.html).
- **3D Reconstruction Experiments**: Various additions, such as transforms, engine changes (e.g., visualization), and the addition of new metrics and losses.
- **Cartesian Subsampling Masks**: Similar to FastMRI but with a twist - selection of the number of center lines instead of center fractions for ACS.
- **Project Organization**: Introduces new project folders for vSHARP and the CMRxRecon challenge.
- **Model Zoo for CMRxRecon Winning Solution**
- **Minor Code Quality Enhancements**

### CODE REVIEW

File: direct/common/subsample.py [hunk 1]
```diff
@@ -324,6 +404,90 @@ def mask_func(
         return torch.from_numpy(self._reshape_and_broadcast_mask(shape, mask))
 
 
+class CartesianEquispacedMaskFunc(FastMRIEquispacedMaskFunc):
+    r"""Cartesian equispaced vertical line mask function.
+
+    Similar to :class:`FastMRIEquispacedMaskFunc, but instead of center fraction (`center_fractions`) representing
+    the fraction of center lines to the original size, here, it represents the actual number of center lines.
+    """
```
Comment (jonasteuwen) [inline direct/common/subsample.py:412]: Perhaps you can sort the classes alphabetically just like you do the export in __init__.py

File: direct/common/subsample.py
Comment (jonasteuwen) [inline direct/common/subsample.py:1]: I would when you edit a file:

- remove # encoding: utf-8 - it is not required in the current python versions. Don't do it everywhere as that will polute the git, but only when you modify a file
- the module docstring of subsample.py is not correct, can you expand it a bit (pylint requires it).

File: direct/data/datasets.py [hunk 2]
```diff
@@ -6,11 +6,13 @@
...
 import pathlib
+import re
 import sys
 import xml.etree.ElementTree as etree  # nosec
 from enum import Enum
 from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
 
+import h5py
 import numpy as np
 from omegaconf import DictConfig
```
Comment (jonasteuwen) [inline direct/data/datasets.py:17]: You can run isort . to sort the imports according to PEP8. Recommended.

File: direct/data/h5_data.py [hunk 3]
```diff
@@ -129,12 +129,13 @@ def __init__(
         else:
             self.logger.info("Using %s h5 files in %s.", len(filenames), self.root)
 
+        self.sensitivity_maps = cast_as_path(sensitivity_maps)
+
```
Comment (jonasteuwen) [inline direct/data/h5_data.py:133]: Can we not start assuming the objects are all pathlib.Path?

File: direct/data/mri_transforms.py [hunk 4]
```diff
@@ -134,6 +136,7 @@ class RandomFlipType(DirectEnum):
     horizontal = "horizontal"
     vertical = "vertical"
     random = "random"
+    both = "both"
```
Comment (jonasteuwen) [inline direct/data/mri_transforms.py:139]: Doesn't conform to UPPERCASE

File: direct/functionals/__init__.py [hunk 5]
```diff
@@ -3,7 +3,9 @@
 
 from direct.functionals.challenges import *
 from direct.functionals.grad import *
+from direct.functionals.hfen import *
 from direct.functionals.nmae import NMAELoss
 from direct.functionals.nmse import *
 from direct.functionals.psnr import *
+from direct.functionals.snr import SNRLoss, snr
 from direct.functionals.ssim import *
```
Comment (jonasteuwen) [inline direct/functionals/__init__.py:11]: I think you need an __all__ if you do this.

File: direct/functionals/hfen.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+#
```
Comment (jonasteuwen) [inline direct/functionals/hfen.py:3]: 1. Remove first line
2. Add module docstring.

File: direct/functionals/hfen.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+import numpy as np
+import torch
+from torch import nn
+
+__all__ = ["hfen_l1", "hfen_l2", "HFENLoss", "HFENL1Loss", "HFENL2Loss"]
+
+
+def get_log_kernel2d(kernel_size: int | list[int] = 5, sigma: Optional[float | list[float]] = None) -> torch.Tensor:
+    """Generates a 2D LoG (Laplacian of Gaussian) kernel.
+
```
Comment (jonasteuwen) [inline direct/functionals/hfen.py:19]: If you do not intend to export, I would propose to prepend the modules with a `_`, this way they don't end up in the documentation and python linters will give a warning when you try to import it from another module.

File: direct/functionals/hfen.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    .. math::
+
+        || \text{LoG}(x_\text{rec}) - \text{LoG}(x_\text{tar}) ||_C
+
+    Where C can be any norm, LoG is the Laplacian of Gaussian filter, and :math:`x_\text{rec}), \text{LoG}(x_\text{tar}`
+    are the reconstructed inp and target images.
+    If normalized it scales it by :math:`|| \text{LoG}(x_\text{tar}) ||_C`.
+
+    Code was borrowed and adapted from _[2].
+
```
Comment (jonasteuwen) [inline direct/functionals/hfen.py:106]: Mention the license of where you derived it, and if it's compatible with the current one.

File: direct/functionals/snr.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+__all__ = ("snr", "SNRLoss")
+
+
+def snr(input_data: torch.Tensor, target_data: torch.Tensor, reduction: str = "mean") -> torch.Tensor:
+    """This function is a torch implementation of SNR metric for batches.
+
+    Parameters
+    ----------
+    input_data : torch.Tensor
```
Comment (jonasteuwen) [inline direct/functionals/snr.py:15]: Definition in the docstring?

File: direct/functionals/ssim.py [hunk 10]
```diff
@@ -60,3 +60,64 @@ def forward(self, X, Y, data_range):
...
+        k2: float
+            k2 parameter for SSIM calculation. Default: 0.03.
+        """
+        super().__init__()
+        self.win_size = win_size
+        self.k1, self.k2 = k1, k2
+
+    def forward(self, X: torch.Tensor, Y: torch.Tensor, data_range: torch.Tensor) -> torch.Tensor:
+        """Forward pass of :class:`SSIM3Dloss`.
+
```
Comment (jonasteuwen) [inline direct/functionals/ssim.py:85]: variables should be snake case, e.g. input and target

File: direct/nn/conv/conv.py [hunk 11]
```diff
@@ -72,3 +74,258 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        weight_mean = weight_.mean(dim=1, keepdim=True)
+        weight_ = weight_ - weight_mean
+        norm = weight_.norm(dim=1, keepdim=True) + 1e-5
+        weight_CWN = weight_ / norm
+        return weight_CWN.view(weight.size())
+
+
+# Custom Conv2d layer with centered weight normalization
+class CWN_Conv2d(nn.Conv2d):
+    """Convolutional layer with Centered Weight Normalization.
```
Comment (jonasteuwen) [inline direct/nn/conv/conv.py:109]: Doesn't conform to PascalCase

File: direct/nn/conv/conv.py [hunk 12]
```diff
@@ -72,3 +74,258 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            The scale factor for the centered weight normalization. Default: 1.414.
+        adjustScale : bool, optional
+            If True, the scale factor is adjusted as a learnable parameter. Default: False.
+        """
+        super().__init__(in_channels, out_channels, kernel_size, stride, padding, dilation, groups, bias, **kwargs)
+        self.weight_normalization = CWNorm()
+        self.scale_ = torch.ones(out_channels, 1, 1, 1).fill_(NScale)
+        if adjustScale:
+            self.WNScale = Parameter(self.scale_)
+        else:
```
Comment (jonasteuwen) [inline direct/nn/conv/conv.py:159]: PascalCase

File: direct/nn/conv/conv.py [hunk 13]
```diff
@@ -72,3 +74,258 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            If True, the layer has a bias term. Default: True.
+        NScale : float, optional
+            The scale factor for the centered weight normalization. Default: 1.414.
+        adjustScale : bool, optional
+            If True, the scale factor is adjusted as a learnable parameter. Default: False.
+        """
+        super().__init__(in_channels, out_channels, kernel_size, stride, padding, dilation, groups, bias, **kwargs)
+        self.weight_normalization = CWNorm()
+        self.scale_ = torch.ones(out_channels, 1, 1, 1).fill_(NScale)
+        if adjustScale:
```
Comment (jonasteuwen) [inline direct/nn/conv/conv.py:157]: snake_case

File: direct/nn/conv/conv.py [hunk 14]
```diff
@@ -72,3 +74,258 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            The output tensor after applying the convolution operation with centered weight normalization.
+        """
+        weight_q = self.weight_normalization(self.weight)
+        weight_q = weight_q * self.WNScale
+        out = F.conv2d(input_f, weight_q, self.bias, self.stride, self.padding, self.dilation, self.groups)
+        return out
+
+
+class CWN_ConvTranspose2d(nn.ConvTranspose2d):
+    """Transposed Convolutional layer with Centered Weight Normalization.
```
Comment (jonasteuwen) [inline direct/nn/conv/conv.py:182]: PascalCase

File: direct/nn/get_nn_model_config.py [hunk 15]
```diff
@@ -11,6 +11,16 @@
...
 
+def _get_activation(activation: ActivationType):
+    return (
+        nn.PReLU()
+        if activation == ActivationType.prelu
+        else nn.ReLU()
+        if activation == ActivationType.relu
+        else nn.LeakyReLU()
+    )
+
```
Comment (jonasteuwen) [inline direct/nn/get_nn_model_config.py:22]: I would do this a bit in steps, this is hard to read

File: direct/nn/mri_models.py [hunk 16]
```diff
@@ -20,7 +21,19 @@
...
+    NMAELoss,
+    NMSELoss,
+    NRMSELoss,
+    PSNRLoss,
+    SNRLoss,
+    SobelGradL1Loss,
+    SobelGradL2Loss,
+    SSIM3DLoss,
+    SSIMLoss,
+)
```
Comment (jonasteuwen) [inline direct/nn/mri_models.py:36]: Maybe it's better to do this `import direct.functionals as D`, and then do `D.SSIMLoss`, etc.

File: direct/utils/__init__.py [hunk 17]
```diff
@@ -232,6 +232,24 @@ def merge_list_of_dicts(list_of_dicts: List[Dict]) -> Dict:
     return functools.reduce(lambda a, b: {**dict(a), **dict(b)}, list_of_dicts)
 
 
+def merge_list_of_lists(list_of_lists: List[List]) -> List:
+    """A list of lists is merged into one list.
```
Comment (jonasteuwen) [inline direct/utils/__init__.py:236]: Your typing needs to be more explicit, so what data types? See `mypy --strict`

File: direct/engine.py [hunk 18]
```diff
@@ -167,6 +167,13 @@ def predict(
         self.ndim = dataset.ndim  # type: ignore
         self.logger.info("Data dimensionality: %s.", self.ndim)
 
+        if self.ndim == 3 and batch_size > 1:
+            self.logger.warning(
+                f"Batch size for inference of 3D data must be 1. Received {batch_size}."
+                f"`batch_size` overwritten by 1."
+            )
+            batch_size = 1
+
```
Comment (jonasteuwen) [inline direct/engine.py:176]: I think you can also do warnings.warn and have it redirected to the logger, that's a bit better.

**Review by jonasteuwen — CHANGES_REQUESTED** [review]
Thank you for the PR. Check my comments.

File: direct/data/datasets.py (hunk 2 above)
Comment (georgeyiasemis) [reply direct/data/datasets.py:17]: This is sorted

Comment (georgeyiasemis) [reply direct/data/datasets.py:17]: isort has been ran for all code. `import` goes first then `from...import..`

File: direct/common/subsample.py (hunk 1 above)
Comment (georgeyiasemis) [reply direct/common/subsample.py:412]: Not possible, `FastMRIEquispacedMaskFunc` for example, is parent of `CartesianEquispacedMaskFunc`

File: direct/functionals/hfen.py (hunk 8 above)
Comment (georgeyiasemis) [reply direct/functionals/hfen.py:106]: There is no license present. I will just mention that (no license exists)

File: direct/data/h5_data.py (hunk 3 above)
Comment (georgeyiasemis) [reply direct/data/h5_data.py:133]: Hmm the above might be tricky as it's an input from the yaml file.

Comment (georgeyiasemis) [reply direct/data/h5_data.py:133]: Btw the above was old code, I just moved it a few lines back.  And I added some more functionality for when sensitivity maps are read from a file.

**Comment by georgeyiasemis** [comment]
> Thank you for the PR. Check my comments.

All checked! I also added readmes for the two new projects.
